exec/java-exec/src/main/java/org/apache/drill/exec/rpc/user/security/HtpasswdFileUserAuthenticator.java

```diff
@@ -0,0 +1,147 @@
+
+  String path = DEFAULT_HTPASSWD_AUTHENTICATOR_PATH;
+  long lastModified;
+  HashMap<String, String> userToPassword;
+
+  @Override
+  public void setup(DrillConfig drillConfig) throws DrillbitStartupException {
+    if (drillConfig.hasPath(ExecConstants.HTPASSWD_AUTHENTICATOR_PATH)) {
+      path = drillConfig.getString(ExecConstants.HTPASSWD_AUTHENTICATOR_PATH);
+    }
+  }
+
+
+  /**
+   * Check password against hash read from the file
+   *
+   * @param password User provided password
+   * @param hash     Hash stored in the htpasswd file
+   * @return true if the password matched the hash
+   */
+  public static boolean checkPassword(String password, String hash) {
+    if (hash.startsWith("$apr1$")) {
+      return hash.equals(Md5Crypt.apr1Crypt(password, hash));
+    } else if (hash.startsWith("$1$")) {
+      return hash.equals(Md5Crypt.md5Crypt(password.getBytes(Charsets.UTF_8), hash));
+    } else if (hash.startsWith("{SHA}")) {
+      return hash.substring(5).equals(Base64.getEncoder().encodeToString(DigestUtils.sha1(password)));
+    } else if (hash.startsWith("$2y$")) {
+      // bcrypt not supported currently
+      return false;
+    } else {
+      return hash.equals(password);
+    }
+  }
+
+  /**
+   * Validate the given username and password against the password file
+   *
+   * @param username Username provided
+   * @param password Password provided
+   * @throws UserAuthenticationException If the username and password could not be validated
+   */
+  @Override
+  public void authenticate(String username, String password) throws UserAuthenticationException {
+    read();
+    String hash = this.userToPassword.get(username);
+    logger.error("Testing " + hash + " against password " + password);
+    boolean credentialsAccepted = (hash != null && this.checkPassword(password, hash));
+    if (!credentialsAccepted) {
+      throw new UserAuthenticationException(String.format("htpasswd auth failed for user '%s'",
+        username));
+    }
+  }
+
+  /**
+   * Read the password file into the map, if the file has changed since we last read it
+   */
+  protected synchronized void read() {
+    File file = new File(path);
+    if (file.exists() && (userToPassword == null || file.lastModified() != lastModified)) {
+      HashMap<String, String> newMap = new HashMap<String, String>();
+      Pattern entry = Pattern.compile("^([^:]+):([^:]+)");
+      try (FileInputStream stream = new FileInputStream(file); Scanner scanner = new Scanner(stream)) {
+        while (scanner.hasNextLine()) {
+          String line = scanner.nextLine().trim();
+          if (!line.isEmpty() && !line.startsWith("#")) {
+            Matcher m = entry.matcher(line);
+            if (m.matches()) {
+              newMap.put(m.group(1), m.group(2));
+            }
+          }
+        }
+      } catch (Exception e) {
+        logger.error(MessageFormat.format("Failed to read htpasswd file at path {0}", file), e);
```

Comment:
Maybe differentiate the two cases: 1) failed to read the file the first time, and so auth will be broken, or 2) failed to read an update; continuing to read the previous version.

Reply:
I think it's better to completely fail in this case, otherwise they won't even realize they messed up the file and won't understand why their changes did not take effect.

Reply:
Might be good to fail if the file can't be read at all, but log an error if the file previously existed, but is now invalid. That way, an error won't take down the cluster.

Of course, in your target use case, Docker, the file won't change within a single run; there is no mechanism to push an updated file to all Drillbits. (You could host-mount the location, and change the file on the host, but this seems to be far too complex and error prone.) The behavior seems fine: to change security, create an updated file and relaunch the container.

An ideal solution would be to allow auth as a plugin to call out to an auth service elsewhere on the cluster. I don't know enough about how that works in practice to suggest how that might work.

Reply:
The logic currently will just log an error if the file fails to load for some reason.  If the file is updated, it could log another error if it still invalid, or it will load it successfully.  It doesn't shut down the drillbit in any case.

I am planning to mount the file from a kubernetes secret, which means I can update it without restarting the drillbits.  It seems like a pain to restart the drillbits after adding/removing a user.  It can take a while to roll all the drillbits and may interrupt running operations and delete query profiles if you do that.